modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -4297,21 +4297,35 @@ void GDScriptAnalyzer::reduce_preload(GDScriptParser::PreloadNode *p_preload) {
 		} else {
 			// TODO: Don't load if validating: use completion cache.
 
-			// Must load GDScript separately to permit cyclic references
-			// as ResourceLoader::load() detects and rejects those.
-			if (ResourceLoader::get_resource_type(p_preload->resolved_path) == "GDScript") {
-				Error err = OK;
-				Ref<GDScript> res = GDScriptCache::get_shallow_script(p_preload->resolved_path, err, parser->script_path);
-				p_preload->resource = res;
-				if (err != OK) {
-					push_error(vformat(R"(Could not preload resource script "%s".)", p_preload->resolved_path), p_preload->path);
-				}
-			} else {
-				p_preload->resource = ResourceLoader::load(p_preload->resolved_path);
-				if (p_preload->resource.is_null()) {
-					push_error(vformat(R"(Could not preload resource file "%s".)", p_preload->resolved_path), p_preload->path);
+			Error err = OK;
+			p_preload->resource = ResourceLoader::load(p_preload->resolved_path, "", ResourceFormatLoader::CACHE_MODE_REUSE, &err);
+			if (err == ERR_BUSY) {
+				// Cyclic reference detected. Handle somehow.
+				const String &res_type = ResourceLoader::get_resource_type(p_preload->resolved_path);
+				if (res_type == "GDScript") {
+					// a) Special handling for GDScript.
+					Error err = OK;
+					Ref<GDScript> res = GDScriptCache::get_shallow_script(p_preload->resolved_path, err, parser->script_path);
+					p_preload->resource = res;
+					if (err != OK) {
+						push_error(vformat(R"(Could not preload resource script "%s".)", p_preload->resolved_path), p_preload->path);
+					}
+				} else {
+					// b) Generic handling for other resource types (get a token and resolve later).
+					Object *obj = ClassDB::instantiate(res_type);
+					p_preload->resource = obj;
+					if (p_preload->resource.is_valid()) {
+						ResourceLoader::set_nested_load_resource_pointer_override(p_preload->resolved_path, p_preload->resource.ptr());
```

Comment:
Regarding if it's a good idea, well, I'm not sure it's the best, either, as it makes tracking the lifetime of the resource harder. 😅 However, that's the heart of this approach. Doing it any later won't work. Also, it's not really fake, but an empty shell until later populated by the relevant resource loader.

Let's go with an example to illustrate my logic:
- Resource loader is asked to load `forest.tscn`.
- Text resource loader starts parsing `forest.tscn`'s external resources.
- One of the ext resources happens to be `forest.gd`.
- The script `forest.gd` happens to preload `forest.tscn`.
- You don't need a fully loaded `forest.tscn` `PackedScene` now. As long as you can point to the resource that will eventualy be that thing once the stack is unrolled, you're fine. So, you construct the object yourself (an empty shell for now, but of the right type) and leave a message to the underlying load call to use the instance you have just created instead of having it create the instance itself. That way, the same resource reference is being used at both places.

Reply:
So, this is the same as Adam's "shallow" resources, with the difference that a resource can be created outside of the `ResourceLoader`?

Reply:
It's the same core idea. In fact, my first attempt was just to implement an equivalent of Adam's PR's `ResourceLoader::get_loading_resource()` in a way that was better scoped to the issue and provide it as a patch to his PR. However, I found in my tests that one can't simply check `thread_load_tasks[local_path].resource` of, say, `forest.tscn`, because it will be null at that point. Therefore, I had to implement the pointer override mechanism and the code ended up diverging a lot. If his PR turns out to work, I'd still like to have that polished and learn what's the secret to `resource` having a valid value when checked, but I'd consider it better than this one.

Regarding memory leaks, yeah, this approach would still need a lot of testing and the resource loading pipeline in the engine may need to see some redesign for a cleaner solution to be possible.

I hope this addresses your concerns.

Reply:
On deeper reflection, rather than memory leaks, there would be a possibility of dangling pointers. However, I'm avoiding that in my next push. Other issues because of edge cases can still happen, though, such as funny interaction with load cache modes, but that's not eclusive of this PR.